File: backend/app/interfaces/schemas/tool_presentation.py
"""Public, bounded schema for declarative tool-result presentation.

The descriptor is data only.  It deliberately cannot name a Vue component,
contain HTML, or carry executable JavaScript.  Tool manifests and tool results
are extension input, so the API boundary normalizes them before they reach an
SSE client.
"""

from __future__ import annotations

import json
from itertools import islice
import math
import re
from typing import Any, Literal
from urllib.parse import parse_qs, urlsplit

from pydantic import BaseModel, ConfigDict, Field


ToolPresentationKind = Literal[
    "auto",
    "generic",
    "table",
    "chart",
    "map",
    "image",
    "artifact",
    "log",
]
ToolChartType = Literal["line", "bar", "area", "scatter"]
ToolLogLevel = Literal["debug", "info", "warning", "error"]

_MAX_DATA_DEPTH = 7
_MAX_COLLECTION_ITEMS = 200
_MAX_OBJECT_FIELDS = 80
_MAX_STRING_LENGTH = 20_000
_MAX_DATA_NODES = 4_000
# This is a byte limit for the complete compact-JSON representation of
# ``ToolPresentation.data``.  Counting only string values is insufficient:
# repeated object keys, quotes, escapes, commas and brackets can otherwise
# make the SSE payload substantially larger than the advertised limit.
MAX_TOOL_PRESENTATION_DATA_BYTES = 256_000
_MAX_COLUMNS = 32
_MAX_SERIES = 16

_SIZE_LIMIT_MARKER = "[content omitted: size limit]"
_NESTING_LIMIT_MARKER = "[content omitted: nesting limit]"

_SECRET_NORMALIZED_KEYS = frozenset({
    "apikey",
    "accesskey",
    "secret",
    "secretkey",
    "clientsecret",
    "password",
    "passwd",
    "credential",
    "authorization",
    "cookie",
    "privatekey",
    "signature",
})
_SECRET_NORMALIZED_SUFFIXES = (
    "apikey",
    "accesskey",
    "secret",
    "secretkey",
    "clientsecret",
    "password",
    "passwd",
    "token",
    "credential",
    "cookie",
    "privatekey",
    "signature",
)
_HOST_PATH = re.compile(
    r"(?<![\w])(?:/(?!home/ubuntu(?:/|\b))[^\s\"'`|;&<>)]*|"
    r"[A-Za-z]:\\[^\s\"'`|;&<>)]*|\\\\[^\s\"'`|;&<>)]*)"
)
_CREDENTIAL_VALUE = re.compile(
    r"(?i)(\b(?:api[-_]?key|access[-_]?key|secret(?:[-_]?key)?|password|"
    r"passwd|(?:[a-z][a-z0-9]*[-_]?)?token|(?:[a-z][a-z0-9]*[-_]?)?credential|"
    r"cookie|(?:[a-z][a-z0-9]*[-_]?)?secret|client[-_]?secret|"
    r"(?:[a-z][a-z0-9]*[-_]?)?password|(?:[a-z][a-z0-9]*[-_]?)?api[-_]?key|"
    r"private[-_]?key|signature)"
    r"\s*[:=]\s*)(?:\"[^\"]*\"|'[^']*'|[^\s,}\]]+)"
)
_AUTHORIZATION_VALUE = re.compile(
    r"(?i)(\bAuthorization\s*[:=]\s*)(Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+"
)
_BARE_AUTH_VALUE = re.compile(
    r"(?i)\b(Bearer|Basic)\s+[A-Za-z0-9._~+/=-]+"
)
_URL_CREDENTIALS = re.compile(
    r"(?i)\b([a-z][a-z0-9+.-]*://)([^\s/@:'\"]+):([^\s/@'\"]+)@"
)
_WEB_SCHEME_MARKER = "\ue000dataseek-web-scheme\ue000"


class ToolPresentationColumn(BaseModel):
    """One safe table column selected by a result object key."""

    model_config = ConfigDict(extra="ignore")

    key: str = Field(min_length=1, max_length=120)
    label: str | None = Field(default=None, max_length=160)
    align: Literal["left", "center", "right"] | None = None


class ToolPresentationSeries(BaseModel):
    """One numeric series selected by a result object key."""

    model_config = ConfigDict(extra="ignore")

    key: str = Field(min_length=1, max_length=120)
    label: str | None = Field(default=None, max_length=160)
    color: str | None = Field(default=None, max_length=32)


class ToolPresentation(BaseModel):
    """Version-one declarative presentation DSL exposed over the existing SSE."""

    model_config = ConfigDict(extra="ignore")

    kind: ToolPresentationKind = "auto"
    title: str | None = Field(default=None, max_length=160)
    description: str | None = Field(default=None, max_length=1_000)
    data: Any | None = None
    columns: list[ToolPresentationColumn] | None = None
    series: list[ToolPresentationSeries] | None = None
    x_key: str | None = Field(default=None, max_length=120)
    chart_type: ToolChartType | None = None
    url: str | None = Field(default=None, max_length=2_048)
    mime_type: str | None = Field(default=None, max_length=160)
    filename: str | None = Field(default=None, max_length=255)
    level: ToolLogLevel | None = None


def _safe_text(value: Any, *, limit: int) -> str | None:
    if not isinstance(value, (str, int, float, bool)):
        return None
    text = str(value).strip()
    if not text:
        return None
    if _safe_resource_url(text) is not None:
        return text[:limit]
    text = _URL_CREDENTIALS.sub(r"\1[redacted credential]@", text)
    text = _AUTHORIZATION_VALUE.sub(r"\1\2 [redacted credential]", text)
    # A bounded head/tail preview can begin in the middle of an Authorization
    # header, so redact a standalone scheme/value pair as well.
    text = _BARE_AUTH_VALUE.sub(r"\1 [redacted credential]", text)
    text = _CREDENTIAL_VALUE.sub(r"\1[redacted credential]", text)
    # Preserve web URL separators while still redacting filesystem paths in
    # surrounding text and URL query values (for example ?path=/Users/...).
    text = text.replace(_WEB_SCHEME_MARKER, "?")
    for scheme in ("https://", "http://", "wss://", "ws://"):
        text = text.replace(scheme, f"{scheme[:-2]}{_WEB_SCHEME_MARKER}")
    text = _HOST_PATH.sub("[protected path]", text)
    text = text.replace(_WEB_SCHEME_MARKER, "//")
    return text[:limit]


def _safe_key(value: Any) -> str | None:
    text = _safe_text(value, limit=120)
    if text is None:
        return None
    normalized = re.sub(r"[^a-z0-9]", "", text.casefold())
    if normalized in _SECRET_NORMALIZED_KEYS or normalized.endswith(
        _SECRET_NORMALIZED_SUFFIXES
    ):
        return None
    return text


def _compact_json_size(value: Any) -> int:
    """Return the exact UTF-8 size used by the public compact JSON contract."""

    return len(json.dumps(
        value,
        ensure_ascii=False,
        allow_nan=False,
        separators=(",", ":"),
    ).encode("utf-8"))


def _fit_json_value(value: Any, max_bytes: int) -> tuple[Any, int] | None:
    """Fit one already-safe scalar, falling back to the size marker."""

    try:
        size = _compact_json_size(value)
    except (TypeError, ValueError, OverflowError):
        value = None
        size = _compact_json_size(value)
    if size <= max_bytes:
        return value, size
    marker_size = _compact_json_size(_SIZE_LIMIT_MARKER)
    if marker_size <= max_bytes:
        return _SIZE_LIMIT_MARKER, marker_size
    return None


def _fit_json_string(value: str, max_bytes: int) -> tuple[Any, int] | None:
    """Return the longest code-point prefix whose JSON encoding fits."""

    full_size = _compact_json_size(value)
    if full_size <= max_bytes:
        return value, full_size
    if _compact_json_size("") > max_bytes:
        return None

    low = 0
    high = len(value)
    while low < high:
        midpoint = (low + high + 1) // 2
        if _compact_json_size(value[:midpoint]) <= max_bytes:
            low = midpoint
        else:
            high = midpoint - 1
    truncated = value[:low]
    return truncated, _compact_json_size(truncated)


def _safe_data_with_size(
    value: Any,
    *,
    depth: int,
    nodes: list[int],
    max_bytes: int,
) -> tuple[Any, int] | None:
    """Sanitize one value and account for every serialized JSON byte."""

    if nodes[0] <= 0:
        return _fit_json_value(_SIZE_LIMIT_MARKER, max_bytes)
    nodes[0] -= 1

    if value is None or isinstance(value, (bool, int)):
        return _fit_json_value(value, max_bytes)
    if isinstance(value, float):
        return _fit_json_value(value if math.isfinite(value) else None, max_bytes)
    if isinstance(value, str):
        safe = _safe_text(value, limit=_MAX_STRING_LENGTH)
        if safe is None:
            return _fit_json_value(None, max_bytes)
        return _fit_json_string(safe, max_bytes)
    if depth >= _MAX_DATA_DEPTH:
        return _fit_json_value(_NESTING_LIMIT_MARKER, max_bytes)
    if hasattr(value, "model_dump"):
        value = value.model_dump(mode="json")
    if isinstance(value, (list, tuple)):
        # The opening and closing brackets are part of the aggregate budget.
        result: list[Any] = []
        serialized_size = 2
        if serialized_size > max_bytes:
            return None
        for item in value[:_MAX_COLLECTION_ITEMS]:
            if nodes[0] <= 0:
                break
            separator_size = 1 if result else 0
            child_budget = max_bytes - serialized_size - separator_size
            child = _safe_data_with_size(
                item,
                depth=depth + 1,
                nodes=nodes,
                max_bytes=child_budget,
            )
            if child is None:
                break
            child_value, child_size = child
            result.append(child_value)
            serialized_size += separator_size + child_size
        return result, serialized_size
    if isinstance(value, dict):
        # The opening/closing braces, serialized keys, colon and commas all
        # count.  This closes the long/repeated-key amplification gap.
        result: dict[str, Any] = {}
        serialized_size = 2
        if serialized_size > max_bytes:
            return None
        for raw_key, item in islice(value.items(), _MAX_OBJECT_FIELDS):
            key = _safe_key(raw_key)
            if key is None or key in result:
                continue
            if nodes[0] <= 0:
                break
            key_size = _compact_json_size(key)
            separator_size = 1 if result else 0
            entry_overhead = separator_size + key_size + 1  # comma + key + colon
            child_budget = max_bytes - serialized_size - entry_overhead
            if child_budget < 0:
                break
            child = _safe_data_with_size(
                item,
                depth=depth + 1,
                nodes=nodes,
                max_bytes=child_budget,
            )
            if child is None:
                break
            child_value, child_size = child
            result[key] = child_value
            serialized_size += entry_overhead + child_size
        return result, serialized_size

    safe = _safe_text(value, limit=_MAX_STRING_LENGTH)
    if safe is None:
        return _fit_json_value(None, max_bytes)
    return _fit_json_string(safe, max_bytes)


def _safe_data(value: Any) -> Any:
    # A single node counter and byte ceiling are shared across the full tree,
    # avoiding exponential expansion and guaranteeing the serialized boundary.
    fitted = _safe_data_with_size(
        value,
        depth=0,
        nodes=[_MAX_DATA_NODES],
        max_bytes=MAX_TOOL_PRESENTATION_DATA_BYTES,
    )
    if fitted is None:
        return _SIZE_LIMIT_MARKER
    safe_value, serialized_size = fitted
    # Keep this assertion next to the encoder contract so later schema changes
    # cannot silently reintroduce character-based accounting.
    assert serialized_size == _compact_json_size(safe_value)
    assert serialized_size <= MAX_TOOL_PRESENTATION_DATA_BYTES
    return safe_value


def sanitize_tool_public_data(value: Any) -> Any:
    """Return one bounded, credential/path-safe value for an SSE payload."""
    return _safe_data(value)


def _safe_resource_url(value: Any) -> str | None:
    """Allow only same-application signed file URLs.

    Relative paths keep the browser on the DataSeek origin.  Scheme-relative,
    external, data, blob and javascript URLs are rejected at the API boundary.
    """

    if not isinstance(value, str) or len(value) > 2_048:
        return None
    parsed = urlsplit(value.strip())
    if parsed.scheme or parsed.netloc or parsed.fragment:
        return None
    if not re.fullmatch(r"/api/v1/files/[A-Za-z0-9._-]{1,255}", parsed.path):
        return None
    if parsed.query:
        query = parse_qs(parsed.query, keep_blank_values=True)
        if set(query) != {"signature", "expires"}:
            return None
        if len(query["signature"]) != 1 or not re.fullmatch(
            r"[a-fA-F0-9]{64}", query["signature"][0]
        ):
            return None
        if len(query["expires"]) != 1 or not re.fullmatch(
            r"[0-9]{1,12}", query["expires"][0]
        ):
            return None
    return value.strip()


def _safe_columns(value: Any) -> list[ToolPresentationColumn] | None:
    if not isinstance(value, list):
        return None
    columns: list[ToolPresentationColumn] = []
    for raw in value[:_MAX_COLUMNS]:
        if isinstance(raw, str):
            raw = {"key": raw}
        if not isinstance(raw, dict):
            continue
        key = _safe_key(raw.get("key"))
        if key is None:
            continue
        align = raw.get("align")
        columns.append(ToolPresentationColumn(
            key=key,
            label=_safe_text(raw.get("label"), limit=160),
            align=(
                align
                if isinstance(align, str) and align in {"left", "center", "right"}
                else None
            ),
        ))
    return columns or None


def _safe_series(value: Any) -> list[ToolPresentationSeries] | None:
    if not isinstance(value, list):
        return None
    series: list[ToolPresentationSeries] = []
    for raw in value[:_MAX_SERIES]:
        if isinstance(raw, str):
            raw = {"key": raw}
        if not isinstance(raw, dict):
            continue
        key = _safe_key(raw.get("key"))
        if key is None:
            continue
        color = _safe_text(raw.get("color"), limit=32)
        if color and not re.fullmatch(
            r"#[0-9a-fA-F]{3,8}|[a-zA-Z]{1,20}|(?:rgb|hsl)a?\([0-9.,% ]+\)",
            color,
        ):
            color = None
        series.append(ToolPresentationSeries(
            key=key,
            label=_safe_text(raw.get("label"), limit=160),
            color=color,
        ))
    return series or None


def normalize_tool_presentation(value: Any) -> ToolPresentation | None:
    """Validate and bound an extension-provided presentation descriptor."""

    if value is None:
        return None
    if hasattr(value, "model_dump"):
        value = value.model_dump(mode="python")
    if not isinstance(value, dict):
        return None

    kind = value.get("kind", "auto")
    if not isinstance(kind, str) or kind not in {
        "auto", "generic", "table", "chart", "map", "image", "artifact", "log",
    }:
        return None

    chart_type = value.get("chart_type")
    level = value.get("level")
    return ToolPresentation(
        kind=kind,
        title=_safe_text(value.get("title"), limit=160),
        description=_safe_text(value.get("description"), limit=1_000),
        data=_safe_data(value.get("data")) if "data" in value else None,
        columns=_safe_columns(value.get("columns")),
        series=_safe_series(value.get("series")),
        x_key=_safe_key(value.get("x_key")),
        chart_type=(
            chart_type
            if isinstance(chart_type, str)
            and chart_type in {"line", "bar", "area", "scatter"}
            else None
        ),
        url=_safe_resource_url(value.get("url")),
        mime_type=_safe_text(value.get("mime_type"), limit=160),
        filename=_safe_text(value.get("filename"), limit=255),
        level=(
            level
            if isinstance(level, str)
            and level in {"debug", "info", "warning", "error"}
            else None
        ),
    )


def _result_payload(value: Any) -> Any:
    """Select the useful public result without exposing runner bookkeeping."""

    if hasattr(value, "model_dump"):
        value = value.model_dump(mode="python")
    if isinstance(value, dict) and "data" in value:
        value = value.get("data")
    if (
        isinstance(value, dict)
        and value.get("status") == "contract_rejected"
    ):
        return None
    if isinstance(value, dict) and "result" in value:
        return value.get("result")
    if isinstance(value, dict) and isinstance(value.get("output"), str):
        output = value["output"].strip()
        if output and len(output.encode("utf-8")) <= MAX_TOOL_PRESENTATION_DATA_BYTES * 2:
            try:
                return json.loads(output)
            except (TypeError, ValueError, json.JSONDecodeError):
                return output
    return value


def _contract_rejected(value: Any) -> bool:
    if hasattr(value, "model_dump"):
        value = value.model_dump(mode="python")
    if isinstance(value, dict) and "data" in value:
        value = value.get("data")
    return isinstance(value, dict) and value.get("status") == "contract_rejected"


def populate_tool_presentation_data(
    presentation: ToolPresentation | None,
    function_result: Any,
) -> ToolPresentation | None:
    """Attach bounded runtime data only when a descriptor asks for a card."""

    if (
        presentation is None
        or presentation.kind == "auto"
        or presentation.data is not None
        or function_result is None
    ):
        return presentation
    if _contract_rejected(function_result):
        return None
    return presentation.model_copy(update={
        "data": _safe_data(_result_payload(function_result)),
    })

File: frontend/src/utils/toolPresentation.ts
import type { ToolContent } from '../types/message';
import type {
  ToolChartType,
  ToolPresentation,
  ToolPresentationColumn,
  ToolPresentationKind,
  ToolPresentationSeries,
} from '../types/toolPresentation';
import { isSensitiveToolDisplayKey, sanitizeToolDisplayText } from './toolDisplay.ts';

const KINDS = new Set<ToolPresentationKind>([
  'auto',
  'generic',
  'table',
  'chart',
  'map',
  'image',
  'artifact',
  'log',
]);
const CHART_TYPES = new Set<ToolChartType>(['line', 'bar', 'area', 'scatter']);
const SAFE_COLOR = /^(?:#[0-9a-f]{3,8}|[a-z]{1,20}|(?:rgb|hsl)a?\([0-9.,% ]+\))$/i;
const MAX_NODES = 4_000;
/** Maximum UTF-8 size of the complete compact-JSON `presentation.data`. */
export const MAX_TOOL_PRESENTATION_DATA_BYTES = 256_000;
const MAX_DEPTH = 7;
const MAX_ITEMS = 200;
const MAX_FIELDS = 80;
const MAX_STRING_LENGTH = 20_000;
const SIZE_LIMIT_MARKER = '[内容过多，已省略]';
const NESTING_LIMIT_MARKER = '[内容层级过深，已省略]';

type DisplayBudget = { nodes: number };
type BoundedValue = { value: unknown; bytes: number };

const utf8Encoder = new TextEncoder();

export interface ResolvedToolPresentation extends ToolPresentation {
  kind: Exclude<ToolPresentationKind, 'auto'>;
  data: unknown;
}

const isRecord = (value: unknown): value is Record<string, unknown> => (
  !!value && typeof value === 'object' && !Array.isArray(value)
);

const safeText = (value: unknown, limit: number): string | undefined => {
  if (typeof value !== 'string' && typeof value !== 'number' && typeof value !== 'boolean') return undefined;
  const text = sanitizeToolDisplayText(value).trim();
  return text ? text.slice(0, limit) : undefined;
};

const safeKey = (value: unknown): string | undefined => {
  const key = safeText(value, 120);
  return key && !isSensitiveToolDisplayKey(key) ? key : undefined;
};

const compactJsonBytes = (value: unknown): number => {
  const serialized = JSON.stringify(value);
  if (serialized === undefined) return Number.POSITIVE_INFINITY;
  return utf8Encoder.encode(serialized).byteLength;
};

const fitJsonValue = (value: unknown, maxBytes: number): BoundedValue | undefined => {
  let safeValue = value;
  let bytes: number;
  try {
    bytes = compactJsonBytes(safeValue);
  } catch {
    safeValue = null;
    bytes = compactJsonBytes(safeValue);
  }
  if (bytes <= maxBytes) return { value: safeValue, bytes };
  const markerBytes = compactJsonBytes(SIZE_LIMIT_MARKER);
  return markerBytes <= maxBytes
    ? { value: SIZE_LIMIT_MARKER, bytes: markerBytes }
    : undefined;
};

const fitJsonString = (value: string, maxBytes: number): BoundedValue | undefined => {
  const fullBytes = compactJsonBytes(value);
  if (fullBytes <= maxBytes) return { value, bytes: fullBytes };
  if (compactJsonBytes('') > maxBytes) return undefined;

  // Array.from iterates Unicode code points, so truncation never creates a
  // broken surrogate pair while finding the longest prefix that fits.
  const codePoints = Array.from(value);
  let low = 0;
  let high = codePoints.length;
  while (low < high) {
    const midpoint = Math.floor((low + high + 1) / 2);
    if (compactJsonBytes(codePoints.slice(0, midpoint).join('')) <= maxBytes) {
      low = midpoint;
    } else {
      high = midpoint - 1;
    }
  }
  const truncated = codePoints.slice(0, low).join('');
  return { value: truncated, bytes: compactJsonBytes(truncated) };
};

const boundedDisplayValueWithSize = (
  value: unknown,
  depth: number,
  budget: DisplayBudget,
  maxBytes: number,
): BoundedValue | undefined => {
  if (budget.nodes <= 0) return fitJsonValue(SIZE_LIMIT_MARKER, maxBytes);
  budget.nodes -= 1;
  if (value === null || value === undefined || typeof value === 'boolean') {
    return fitJsonValue(value === undefined ? null : value, maxBytes);
  }
  if (typeof value === 'number') {
    return fitJsonValue(Number.isFinite(value) ? value : null, maxBytes);
  }
  if (typeof value === 'string') {
    const text = sanitizeToolDisplayText(value).slice(0, MAX_STRING_LENGTH);
    return fitJsonString(text, maxBytes);
  }
  if (depth >= MAX_DEPTH) return fitJsonValue(NESTING_LIMIT_MARKER, maxBytes);
  if (Array.isArray(value)) {
    const result: unknown[] = [];
    let bytes = 2; // opening and closing brackets
    if (bytes > maxBytes) return undefined;
    for (const item of value.slice(0, MAX_ITEMS)) {
      if (budget.nodes <= 0) break;
      const separatorBytes = result.length ? 1 : 0;
      const child = boundedDisplayValueWithSize(
        item,
        depth + 1,
        budget,
        maxBytes - bytes - separatorBytes,
      );
      if (!child) break;
      result.push(child.value);
      bytes += separatorBytes + child.bytes;
    }
    return { value: result, bytes };
  }
  if (isRecord(value)) {
    const result: Record<string, unknown> = {};
    let bytes = 2; // opening and closing braces
    if (bytes > maxBytes) return undefined;
    for (const [rawKey, item] of Object.entries(value).slice(0, MAX_FIELDS)) {
      if (budget.nodes <= 0) break;
      const key = safeKey(rawKey);
      if (!key || Object.prototype.hasOwnProperty.call(result, key)) continue;
      const separatorBytes = Object.keys(result).length ? 1 : 0;
      const entryOverhead = separatorBytes + compactJsonBytes(key) + 1; // comma + key + colon
      const childBudget = maxBytes - bytes - entryOverhead;
      if (childBudget < 0) break;
      const child = boundedDisplayValueWithSize(
        item,
        depth + 1,
        budget,
        childBudget,
      );
      if (!child) break;
      result[key] = child.value;
      bytes += entryOverhead + child.bytes;
    }
    return { value: result, bytes };
  }
  const text = safeText(value, MAX_STRING_LENGTH);
  return text === undefined
    ? fitJsonValue(null, maxBytes)
    : fitJsonString(text, maxBytes);
};

const boundedDisplayValue = (value: unknown): unknown => {
  const fitted = boundedDisplayValueWithSize(
    value,
    0,
    { nodes: MAX_NODES },
    MAX_TOOL_PRESENTATION_DATA_BYTES,
  );
  if (!fitted) return SIZE_LIMIT_MARKER;
  // This final guard is deliberately based on the actual browser serializer.
  // It remains fail-closed if future value types change the accounting rules.
  return compactJsonBytes(fitted.value) <= MAX_TOOL_PRESENTATION_DATA_BYTES
    ? fitted.value
    : SIZE_LIMIT_MARKER;
};

const normalizeColumns = (value: unknown): ToolPresentationColumn[] | undefined => {
  if (!Array.isArray(value)) return undefined;
  const columns: ToolPresentationColumn[] = [];
  for (const raw of value.slice(0, 32)) {
    const source = typeof raw === 'string' ? { key: raw } : raw;
    if (!isRecord(source)) continue;
    const key = safeKey(source.key);
    if (!key) continue;
    const align = source.align;
    columns.push({
      key,
      label: safeText(source.label, 160),
      align: align === 'left' || align === 'center' || align === 'right' ? align : undefined,
    });
  }
  return columns.length ? columns : undefined;
};

const normalizeSeries = (value: unknown): ToolPresentationSeries[] | undefined => {
  if (!Array.isArray(value)) return undefined;
  const series: ToolPresentationSeries[] = [];
  for (const raw of value.slice(0, 16)) {
    const source = typeof raw === 'string' ? { key: raw } : raw;
    if (!isRecord(source)) continue;
    const key = safeKey(source.key);
    if (!key) continue;
    const color = safeText(source.color, 32);
    series.push({
      key,
      label: safeText(source.label, 160),
      color: color && SAFE_COLOR.test(color) ? color : undefined,
    });
  }
  return series.length ? series : undefined;
};

export const safeToolResourceUrl = (value: unknown, origin?: string): string | undefined => {
  if (typeof value !== 'string' || value.length > 2_048) return undefined;
  const raw = value.trim();
  if (!raw.startsWith('/') || raw.startsWith('//')) return undefined;
  try {
    const base = origin || 'http://dataseek.local';
    const parsed = new URL(raw, base);
    if (parsed.origin !== new URL(base).origin) return undefined;
    if (parsed.username || parsed.password) return undefined;
    if (parsed.hash) return undefined;
    if (!/^\/api\/v1\/files\/[A-Za-z0-9._-]{1,255}$/.test(parsed.pathname)) return undefined;
    const query = [...parsed.searchParams.entries()];
    if (query.length) {
      if (query.length !== 2) return undefined;
      const signatures = parsed.searchParams.getAll('signature');
      const expirations = parsed.searchParams.getAll('expires');
      if (signatures.length !== 1 || expirations.length !== 1) return undefined;
      if (!/^[0-9a-f]{64}$/i.test(signatures[0])) return undefined;
      if (!/^\d{1,12}$/.test(expirations[0])) return undefined;
    }
    return raw;
  } catch {
    return undefined;
  }
};

const resultPayload = (tool: ToolContent): unknown => {
  if (tool.presentation?.data !== undefined && tool.presentation.data !== null) {
    return tool.presentation.data;
  }
  let value: unknown = tool.content;
  if (isRecord(value) && 'result' in value) value = value.result;
  if (isRecord(value) && 'data' in value && (
    'success' in value || 'message' in value || Object.keys(value).length === 1
  )) {
    value = value.data;
  }
  return value;
};

/** Validate a descriptor again at the browser boundary and attach safe result data. */
export const resolveToolPresentation = (tool: ToolContent): ResolvedToolPresentation | null => {
  const raw = tool.presentation;
  if (!raw || !KINDS.has(raw.kind) || raw.kind === 'auto') return null;
  const chartType = raw.chart_type;
  const level = raw.level;
  return {
    kind: raw.kind,
    title: safeText(raw.title, 160),
    description: safeText(raw.description, 1_000),
    data: boundedDisplayValue(resultPayload(tool)),
    columns: normalizeColumns(raw.columns),
    series: normalizeSeries(raw.series),
    x_key: safeKey(raw.x_key),
    chart_type: chartType && CHART_TYPES.has(chartType) ? chartType : undefined,
    url: safeToolResourceUrl(raw.url),
    mime_type: safeText(raw.mime_type, 160),
    filename: safeText(raw.filename, 255),
    level: level === 'debug' || level === 'info' || level === 'warning' || level === 'error'
      ? level
      : undefined,
  };
};

export const presentationRows = (data: unknown): Record<string, unknown>[] => {
  const candidates: unknown[] = [data];
  if (isRecord(data)) {
    candidates.push(data.rows, data.preview, data.records, data.items, data.points, data.data);
    for (const nested of [data.summary, data.result]) {
      if (!isRecord(nested)) continue;
      candidates.push(
        nested.rows,
        nested.preview,
        nested.records,
        nested.items,
        nested.points,
        nested.data,
      );
    }
  }
  const value = candidates.find(Array.isArray);
  return Array.isArray(value) ? value.filter(isRecord).slice(0, 100) : [];
};

export const presentationColumns = (
  presentation: ResolvedToolPresentation,
  rows: Record<string, unknown>[],
): ToolPresentationColumn[] => {
  if (presentation.columns?.length) return presentation.columns.slice(0, 16);
  const keys: string[] = [];
  for (const row of rows.slice(0, 20)) {
    for (const key of Object.keys(row)) {
      if (!keys.includes(key) && !isSensitiveToolDisplayKey(key)) keys.push(key);
      if (keys.length >= 12) return keys.map((item) => ({ key: item, label: item }));
    }
  }
  return keys.map((item) => ({ key: item, label: item }));
};

export const presentationText = (value: unknown, maxLength = 30_000): string => {
  if (typeof value === 'string') return sanitizeToolDisplayText(value).slice(0, maxLength);
  try {
    return JSON.stringify(value, null, 2).slice(0, maxLength);
  } catch {
    return sanitizeToolDisplayText(value).slice(0, maxLength);
  }
};

export interface GeoPresentationSummary {
  featureCount: number;
  geometryTypes: string[];
  bounds?: [number, number, number, number];
  points: Array<[number, number]>;
}

export const summarizeGeoPresentation = (data: unknown): GeoPresentationSummary => {
  const points: Array<[number, number]> = [];
  const geometryTypes = new Set<string>();
  let featureCount = 0;

  const visitCoordinates = (value: unknown, depth = 0) => {
    if (!Array.isArray(value) || depth > 8 || points.length >= 1_000) return;
    if (
      value.length >= 2
      && typeof value[0] === 'number'
      && Number.isFinite(value[0])
      && typeof value[1] === 'number'
      && Number.isFinite(value[1])
    ) {
      points.push([value[0], value[1]]);
      return;
    }
    for (const item of value.slice(0, 1_000)) visitCoordinates(item, depth + 1);
  };

  const visitGeometry = (value: unknown) => {
    if (!isRecord(value)) return;
    if (typeof value.type === 'string') geometryTypes.add(value.type.slice(0, 80));
    visitCoordinates(value.coordinates);
    if (Array.isArray(value.geometries)) value.geometries.slice(0, 200).forEach(visitGeometry);
  };

  if (isRecord(data) && data.type === 'FeatureCollection' && Array.isArray(data.features)) {
    featureCount = data.features.length;
    data.features.slice(0, 1_000).forEach((feature) => {
      if (isRecord(feature)) visitGeometry(feature.geometry);
    });
  } else if (isRecord(data) && data.type === 'Feature') {
    featureCount = 1;
    visitGeometry(data.geometry);
  } else {
    visitGeometry(data);
  }

  const xs = points.map((point) => point[0]);
  const ys = points.map((point) => point[1]);
  const bounds = points.length
    ? [Math.min(...xs), Math.min(...ys), Math.max(...xs), Math.max(...ys)] as [number, number, number, number]
    : undefined;
  return { featureCount, geometryTypes: [...geometryTypes], bounds, points };
};
